src/arith/rewrite_simplify.cc

```diff
@@ -1557,7 +1591,44 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const NotNode* op) {
 }
 
 PrimExpr RewriteSimplifier::Impl::VisitExpr_(const AndNode* op) {
-  PrimExpr ret = IRMutatorWithAnalyzer::VisitExpr_(op);
+  PrimExpr ret = [&]() -> PrimExpr {
+    // If this extension isn't enabled, just delegate out.
+    if (!(enabled_extensions_ & kApplyConstraintsToBooleanBranches)) {
+      return IRMutatorWithAnalyzer::VisitExpr_(op);
+    }
+
+    PrimExpr a = op->a;
+    PrimExpr b = op->b;
+
+    // Alternate which branch is used as the constraint, and which
+    // is being simplified.  Because some sub-analyzers expect their
+    // constraints to already be simplified, each branch may require
+    // more than update.  The loop condition allows each branch to be
+    // visited up to twice, but only if performs the second visit if
+    // necessary.
+    for (size_t i = 0, last_updated = 0; i < last_updated + 2 && i < 4; i++) {
```

Comment:
nit: Is there a way to make this for-loop more understandable? The multiple iterator makes it difficult to understand, possibly error prone, without detailed study.

Reply:
Yeah, I wasn't sure the best way to express this one.  I only wanted to perform one update per loop iteration, to avoid the fourth simplify (a -> b -> a -> **b**) if it converged earlier.

Hmm, maybe if I rewrite it as a fixed-length loop with early bail-out instead.  I'd initially written it without the `i < 4`, so that it would iterate until convergence, but now that I know how the maximum number of loops required, that flexibility isn't needed anymore.

Reply:
Rewritten, and it definitely is more coherent this way.

Reply:
And updated in the OrNode handler as well.